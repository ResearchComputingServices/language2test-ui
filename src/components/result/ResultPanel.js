import React from 'react';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import PropTypes from 'prop-types';
import Card from '@material-ui/core/Card';
import ResultHeader from './ResultHeader';

function ResultPanel({
    title,
    marks,
    pending,
    defaultExpanded,
    children,
}) {
    return (
        children
            ? (
                <ExpansionPanel
                    className='results-result-panel'
                    defaultExpanded={defaultExpanded}
                >
                    <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
                        <ResultHeader
                            marks={marks}
                            pending={pending}
                            size='medium'
                            title={title}
                        />
                    </ExpansionPanelSummary>
                    <ExpansionPanelDetails>
                        <div className='results-result-panel-details'>
                            {children}
                        </div>
                    </ExpansionPanelDetails>
                </ExpansionPanel>
            )
            : (
                <Card className='results-result-panel-non-expandable'>
                    <ResultHeader
                        marks={marks}
                        pending={pending}
                        size='medium'
                        title={title}
                    />
                </Card>
            )
    );
}

ResultPanel.propTypes = {
    title: PropTypes.string.isRequired,
    marks: PropTypes.string,
    defaultExpanded: PropTypes.bool,
    children: PropTypes.node,
    pending: PropTypes.bool,
};

ResultPanel.defaultProps = {
    marks: null,
    defaultExpanded: false,
    children: null,
    pending: false,
};

export default ResultPanel;
