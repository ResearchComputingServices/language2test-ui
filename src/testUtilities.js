/* eslint-disable import/prefer-default-export */
/* eslint-disable prefer-const */
/* eslint-disable react/prop-types */
/* eslint-disable react/no-children-prop */
import React from 'react';
import { shallow, mount as mountComponent } from 'enzyme';
import { Router } from 'react-router-dom';
import toJSON from 'enzyme-to-json';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import _ from 'lodash';
import historyService from './services/HistoryService';
import { reducers } from './redux/slices';

const mockStore = configureMockStore({ reducers });

function WrapperComponent({ propsFunc, children, Component }) {
    const props = propsFunc();
    return (
        <Component
            children={children}
            {...props}
        />
    );
}

export function createTest(Component, opts = {}, callback) {
    let { props, children, mount, state } = opts;
    if (_.isNil(props)) props = {};
    if (_.isNil(children)) children = null;
    if (mount) {
        return () => {
            state = state || {};
            const mockMath = Object.create(global.Math);
            mockMath.random = () => 0.5;
            global.Math = mockMath;
            const mountedTree = mountComponent((
                <Router history={historyService.getHistory()}>
                    <Provider store={mockStore({ ...state })}>
                        <Component {...props} >
                            {children}
                        </Component>
                    </Provider>
                </Router>));
            _.isFunction(callback) && callback(mountedTree);
        };
    }
    if (_.isFunction(props)) {
        return () => {
            const shallowTree = shallow(
                <WrapperComponent
                    Component={Component}
                    propsFunc={props}
                >
                    {children}
                </WrapperComponent>,
            ).dive();
            _.isFunction(callback) && callback(shallowTree);
        };
    }
    return () => {
        const shallowTree = shallow((
            <Component {...props} >
                {children}
            </Component>
        ));
        _.isFunction(callback) && callback(shallowTree);
    };
}

export function createSnapshotTest(Component, opts = {}) {
    return createTest(Component, opts, tree => expect(toJSON(tree)).toMatchSnapshot());
}
