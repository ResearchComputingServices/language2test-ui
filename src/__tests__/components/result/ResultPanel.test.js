import React from 'react';
import ResultPanel from '../../../components/result/ResultPanel';
import { createSnapshotTest } from '../../../testUtilities';

describe('ResultPanel', () => {
    it('should match snapshot', createSnapshotTest(ResultPanel, {
        props: {
            title: 'Some Title',
            marks: '0/100',
        },
    }), <div>Test Child</div>);
});
