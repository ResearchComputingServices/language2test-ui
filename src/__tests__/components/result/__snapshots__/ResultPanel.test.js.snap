// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultPanel should match snapshot 1`] = `
<WithStyles(ForwardRef(Card))
  className="results-result-panel-non-expandable"
>
  <ResultHeader
    className=""
    marks="0/100"
    pending={false}
    size="medium"
    title="Some Title"
  />
</WithStyles(ForwardRef(Card))>
`;
